wagtail/wagtailsearch/backends/elasticsearch.py

```diff
@@ -432,40 +432,101 @@ def _get_es_body(self, for_count=False):
+
+                # Find objects in database and add them to dict
+                queryset = self.query.queryset.filter(pk__in=pks)
+                for obj in queryset:
+                    results[str(obj.pk)] = obj
+
+                    if self._score_field:
+                        setattr(obj, self._score_field, scores.get(str(obj.pk)))
+
+                # Yield results in order given by Elasticsearch
+                for pk in pks:
+                    if results[str(pk)]:
+                        yield results[str(pk)]
+
+                        if limit is not None:
+                            limit -= 1
+
+                            if limit == 0:
+                                break
+
+                if limit is not None and limit == 0:
+                    break
+
+                # Fetch next page of results
+                if '_scroll_id' not in page:
+                    break
+
+                page = self.backend.es.scroll(scroll_id=page['_scroll_id'], scroll='2m')
+
+            # Clear the scroll
+            if '_scroll_id' in page:
+                self.backend.es.clear_scroll(scroll_id=page['_scroll_id'])
+        else:
+            params.update({
+                'size': limit or PAGE_SIZE,
+            })
 
-        # Send to Elasticsearch
-        hits = self.backend.es.search(**params)
+            # Send to Elasticsearch
+            hits = self.backend.es.search(**params)
 
-        # Get pks from results
-        pks = [hit['fields']['pk'][0] for hit in hits['hits']['hits']]
-        scores = {str(hit['fields']['pk'][0]): hit['_score'] for hit in hits['hits']['hits']}
+            # Get pks from results
+            pks = [hit['fields']['pk'][0] for hit in hits['hits']['hits']]
+            scores = {str(hit['fields']['pk'][0]): hit['_score'] for hit in hits['hits']['hits']}
 
-        # Initialise results dictionary
-        results = dict((str(pk), None) for pk in pks)
+            # Initialise results dictionary
+            results = dict((str(pk), None) for pk in pks)
 
-        # Find objects in database and add them to dict
-        queryset = self.query.queryset.filter(pk__in=pks)
-        for obj in queryset:
-            results[str(obj.pk)] = obj
+            # Find objects in database and add them to dict
+            queryset = self.query.queryset.filter(pk__in=pks)
+            for obj in queryset:
+                results[str(obj.pk)] = obj
 
-            if self._score_field:
-                setattr(obj, self._score_field, scores.get(str(obj.pk)))
+                if self._score_field:
+                    setattr(obj, self._score_field, scores.get(str(obj.pk)))
 
-        # Return results in order given by Elasticsearch
-        return [results[str(pk)] for pk in pks if results[str(pk)]]
+            # Yield results in order given by Elasticsearch
+            for pk in pks:
+                if results[str(pk)]:
+                    yield results[str(pk)]
```

Comment:
Define a variable `result = results[str(pk)]`. This avoids an extra string conversion & and dictionary lookup in almost all cases.